app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,544 @@
+import org.kiwix.kiwixmobile.base.BaseActivity;
+import org.kiwix.kiwixmobile.main.MainActivity;
+import org.kiwix.kiwixmobile.utils.AlertDialogShower;
+import org.kiwix.kiwixmobile.utils.KiwixDialog;
+import org.kiwix.kiwixmobile.wifi_hotspot.HotspotService;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.SelectionMode;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BookOnDiskDelegate;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BooksOnDiskAdapter;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BooksOnDiskListItem;
+
+import static org.kiwix.kiwixmobile.utils.StyleUtils.dialogStyle;
+import static org.kiwix.kiwixmobile.webserver.WebServerHelper.getAddress;
+import static org.kiwix.kiwixmobile.webserver.WebServerHelper.isServerStarted;
+
+public class ZimHostActivity extends BaseActivity implements
+    ServerStateListener, ZimHostContract.View {
+
+  @BindView(R.id.startServerButton)
+  Button startServerButton;
+  @BindView(R.id.server_textView)
+  TextView serverTextView;
+
+  @Inject
+  ZimHostContract.Presenter presenter;
+
+  @Inject AlertDialogShower alertDialogShower;
+
+  public static final String ACTION_TURN_ON_AFTER_O = "Turn_on_hotspot_after_oreo";
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  public static final String ACTION_IS_HOTSPOT_ENABLED = "Is_hotspot_enabled";
+  public static final String ACTION_START_SERVER = "start_server";
+  public static final String ACTION_STOP_SERVER = "stop_server";
+  public static final String SELECTED_ZIM_PATHS_KEY = "selected_zim_paths";
+  private static final String IP_STATE_KEY = "ip_state_key";
+  private static final String TAG = "ZimHostActivity";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  private static final int LOCATION_SETTINGS_PERMISSION_RESULT = 101;
+  private Intent serviceIntent;
+  private Task<LocationSettingsResponse> task;
+
+  private BooksOnDiskAdapter booksAdapter;
+  HotspotService hotspotService;
+  private String ip;
+  boolean bound;
+  private ServiceConnection serviceConnection;
+  private ArrayList<String> selectedBooksPath = new ArrayList<>();
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    AndroidInjection.inject(this);
+
+    setUpToolbar();
+
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(IP_STATE_KEY);
+      serverTextView.setText(
+          getString(R.string.server_started_message, ip));
+      startServerButton.setText(getString(R.string.stop_server_label));
+      startServerButton.setBackgroundColor(getResources().getColor(R.color.stopServer));
+    }
+    BookOnDiskDelegate.BookDelegate bookDelegate =
+        new BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil,
+            bookOnDiskItem -> {
+              open(bookOnDiskItem);
```

Comment:
this will open the book? We are never in the mode where that is possible but set this to null and delete the method

Reply:
Fixed. cd74150f214bff6aa7ce577469c18791d71d8baf

Reply:
Also added null checks. 301368ecbd3fb69a69c267facccb9c85fcfc5749